File: scripts/build_mmdb.go
package main

import (
    "bufio"
    "encoding/json"
    "fmt"
    "log"
    "net"
    "os"
    "path/filepath"
    "strings"

    "github.com/maxmind/mmdbwriter"
    "github.com/maxmind/mmdbwriter/mmdbtype"
)

const (
    dataDir    = "data"
    outputMMDB = "GeoCN.mmdb"

    ipv4Src = "ipv4_source.txt"
    ipv6Src = "ipv6_source.txt"
)

// -------------------- 行政区划结构 --------------------
type County struct {
    Label string `json:"label"`
    Name  string `json:"name"`
    Code  string `json:"code"`
}

type City struct {
    Label    string   `json:"label"`
    Name     string   `json:"name"`
    Code     string   `json:"code"`
    Counties []County `json:"counties"`
}

type Province struct {
    Label  string `json:"label"`
    Name   string `json:"name"`
    Code   string `json:"code"`
    Cities []City `json:"cities"`
}

var provinces []Province

func trimZero(s string) string {
    for len(s) > 0 && s[len(s)-1] == '0' {
        s = s[:len(s)-1]
    }
    return s
}

// -------------------- 三级行政区划查询 --------------------
func findCodes(province, city, district string) (int, int, int) {
    pCode, cCode, dCode := 0, 0, 0

    for _, p := range provinces {
        if p.Name == province {
            pCode = atoi(trimZero(p.Code))

            for _, c := range p.Cities {
                if c.Name == city {
                    cCode = atoi(trimZero(c.Code))

                    for _, d := range c.Counties {
                        if d.Name == district {
                            dCode = atoi(trimZero(d.Code))
                            return pCode, cCode, dCode
                        }
                    }
                }
            }
        }
    }
    return pCode, cCode, dCode
}

func atoi(s string) int {
    n := 0
    for _, c := range s {
        n = n*10 + int(c-'0')
    }
    return n
}

// -------------------- 新增：坐标表 --------------------
var coordsRaw []map[string][2]float64
var coords map[string][2]float64

//
// -------------------- 原有结构体（保持不变） --------------------
//
type Record struct {
    ISP           string
    Net           string
    Province      string
    City          string
    Districts     string
    ProvinceCode  int
    CityCode      int
    DistrictsCode int

    // 新增字段
    Lat float64
    Lng float64
}

// 按样本字段顺序解析：
// startIP|endIP|...|province|city|districts|isp|net|provinceCode|cityCode|districtsCode
func parseLine(line string) (string, string, Record, bool) {
    parts := strings.Split(strings.TrimSpace(line), "|")
    if len(parts) < 9 {
        return "", "", Record{}, false
    }

    return parts[0], parts[1], Record{
        Province:      parts[4],
        City:          parts[5],
        Districts:     parts[6],
        ISP:           parts[7],
        Net:           parts[8],
        ProvinceCode:  0,
        CityCode:      0,
        DistrictsCode: 0,
    }, true
}

// 输出字段严格等于样板
func toMMDBRecord(r Record) mmdbtype.DataType {
    return mmdbtype.Map{
        "isp":           mmdbtype.String(r.ISP),
        "net":           mmdbtype.String(r.Net),
        "province":      mmdbtype.String(r.Province),
        "city":          mmdbtype.String(r.City),
        "districts":     mmdbtype.String(r.Districts),
        "provinceCode":  mmdbtype.Int32(r.ProvinceCode),
        "cityCode":      mmdbtype.Int32(r.CityCode),
        "districtsCode": mmdbtype.Int32(r.DistrictsCode),

        // 新增
        "lat": mmdbtype.Float64(r.Lat),
        "lng": mmdbtype.Float64(r.Lng),
    }
}

// -------------------- 新增：行政名称 → coords key --------------------
func trimSuffix(s string) string {
    if strings.HasSuffix(s, "省") {
        return strings.TrimSuffix(s, "省")
    }
    if strings.HasSuffix(s, "市") {
        return strings.TrimSuffix(s, "市")
    }
    if strings.HasSuffix(s, "区") {
        return strings.TrimSuffix(s, "区")
    }
    if strings.HasSuffix(s, "县") {
        return strings.TrimSuffix(s, "县")
    }
    return s
}

func buildCoordKey(province, city, district string) string {
    return trimSuffix(province) + trimSuffix(city) + trimSuffix(district)
}

func processFile(writer *mmdbwriter.Tree, filePath string) {
    f, err := os.Open(filePath)
    if err != nil {
        log.Printf("skip missing file: %s", filePath)
        return
    }
    defer f.Close()

    scanner := bufio.NewScanner(f)
    for scanner.Scan() {
        start, end, record, ok := parseLine(scanner.Text())
        if !ok {
            continue
        }

        //
        // ----------- 根据省/市/区县查询行政代码 -----------
        //
        p, c, d := findCodes(record.Province, record.City, record.Districts)
        record.ProvinceCode = p
        record.CityCode = c
        record.DistrictsCode = d

        // ----------- 新增：坐标匹配（O(1））-----------
        key := buildCoordKey(record.Province, record.City, record.Districts)
        if v, ok := coords[key]; ok {
            record.Lng = v[0]
            record.Lat = v[1]
        }

        startIP := net.ParseIP(start)
        endIP := net.ParseIP(end)
        if startIP == nil || endIP == nil {
            continue
        }

        writer.InsertRange(startIP, endIP, toMMDBRecord(record))
    }
}

func main() {
    //
    // ----------- 加载行政区划 JSON -----------
    //
    b, err := os.ReadFile("location.json")
    if err != nil {
        log.Fatalf("load location.json error: %v", err)
    }
    if err := json.Unmarshal(b, &provinces); err != nil {
        log.Fatalf("parse location.json error: %v", err)
    }

    // ----------- 加载 coords.json（数组） -----------
    cb, err := os.ReadFile("coords.json")
    if err != nil {
        log.Fatalf("load coords.json error: %v", err)
    }
    if err := json.Unmarshal(cb, &coordsRaw); err != nil {
        log.Fatalf("parse coords.json error: %v", err)
    }

    // ----------- 转换为 map（恢复 O(1) 性能） -----------
    coords = make(map[string][2]float64)
    for _, m := range coordsRaw {
        for k, v := range m {
            coords[k] = v
        }
    }

    outputPath := filepath.Join(dataDir, outputMMDB)
    fmt.Println("Building MMDB:", outputPath)

    writer, err := mmdbwriter.New(mmdbwriter.Options{
        DatabaseType: "GeoCN",
        Languages:    []string{"zh-CN"},
        Description:  map[string]string{"zh-CN": "GeoCN mmdb"},
    })
    if err != nil {
        log.Fatalf("writer init error: %v", err)
    }

    processFile(writer, filepath.Join(dataDir, ipv4Src))
    processFile(writer, filepath.Join(dataDir, ipv6Src))

    f, err := os.Create(outputPath)
    if err != nil {
        log.Fatalf("file create error: %v", err)
    }
    defer f.Close()

    _, err = writer.WriteTo(f)
    if err != nil {
        log.Fatalf("write mmdb error: %v", err)
    }

    fmt.Println("MMDB build completed:", outputPath)
}

File: scripts/build_mmdb.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import maxminddb.writer
import ipaddress

DATA_DIR = "data"
OUTPUT_FILE = os.path.join(DATA_DIR, "ip2region.mmdb")

# 输入文件
IPV4_SRC = os.path.join(DATA_DIR, "ipv4_source.txt")
IPV6_SRC = os.path.join(DATA_DIR, "ipv6_source.txt")


def parse_line(line):
    """
    解析一行：
    1.0.1.0|1.0.3.255|中国|福建省|福州市|电信|AS4134
    """
    parts = line.strip().split("|")
    if len(parts) < 7:
        return None

    start_ip = parts[0]
    end_ip = parts[1]

    record = {
        "country": parts[2],
        "region": parts[3],
        "city": parts[4],
        "isp": parts[5],
        "asn": parts[6],
    }

    return start_ip, end_ip, record


def cidr_merge(start_ip, end_ip):
    """
    将起止 IP 转成 CIDR 列表
    """
    start = int(ipaddress.ip_address(start_ip))
    end = int(ipaddress.ip_address(end_ip))

    cidrs = []
    while start <= end:
        max_size = start & -start
        max_len = (end - start) + 1
        size = max_size if max_size <= max_len else max_len
        cidrs.append(ipaddress.ip_network((start, size), strict=False))
        start += size

    return cidrs


def build_mmdb():
    print("Building mmdb:", OUTPUT_FILE)

    writer = maxminddb.writer.Writer(
        OUTPUT_FILE,
        ip_version=6,
        record_size=32,
        database_type="ip2region",
        languages=["zh-CN"],
        description={"zh-CN": "ip2region mmdb"},
    )

    # 处理 IPv4
    if os.path.exists(IPV4_SRC):
        with open(IPV4_SRC, "r", encoding="utf-8") as f:
            for line in f:
                parsed = parse_line(line)
                if not parsed:
                    continue
                start_ip, end_ip, record = parsed
                for cidr in cidr_merge(start_ip, end_ip):
                    writer.insert_network(cidr, record)

    # 处理 IPv6
    if os.path.exists(IPV6_SRC):
        with open(IPV6_SRC, "r", encoding="utf-8") as f:
            for line in f:
                parsed = parse_line(line)
                if not parsed:
                    continue
                start_ip, end_ip, record = parsed
                for cidr in cidr_merge(start_ip, end_ip):
                    writer.insert_network(cidr, record)

    writer.close()
    print("MMDB build completed:", OUTPUT_FILE)


if __name__ == "__main__":
    build_mmdb()
